kubelet: don't GC containers of deleted pods until all sources are ready

yujuhong opened the pull request:
Otherwise, kubelet may mistaken a pod being deleted and remove all its
containers, leading to potentially incorrect interpretation of the restart
policy (since kubelet sometimes relies on exited containers as checkpoints)

This fixes #27095

yujuhong commented:
I don't like adding more dependencies between components. Suggestions are welcome.

Random-Liu commented:
It seems that it won't be too hard to move `StartGarbageCollection` into kubelet.Run.
Just call `StartGarbageCollection` into kubelet.Run(), and add a new layer of starter goroutine checking source periodically and start garbage collector when all sources are ready. The change should also be minor.

yujuhong commented:
Alternative solution:
```
func (kl *Kubelet) StartGarbageCollection() {
        go wait.Until(func() {
                if kl.sourcesReady.AllReady() {
                        if err := kl.containerGC.GarbageCollect(); err != nil {
                                glog.Errorf("Container garbage collection failed: %v", err)
                        }
                }
        }, ContainerGCPeriod, wait.NeverStop)

        go wait.Until(func() {
                if err := kl.imageManager.GarbageCollect(); err != nil {
                        glog.Errorf("Image garbage collection failed: %v", err)
                }
        }, ImageGCPeriod, wait.NeverStop)
}
```
Pros: 2-line change
Cons: This blocks the entire container garbage collection until all sources are ready.

Another option is to pass a new argument to indicate whether the sources area ready:
```
                        if err := kl.containerGC.GarbageCollect(kl.sourcesReady.AllReady()); err != nil {
                                glog.Errorf("Container garbage collection failed: %v", err)
                        }
```
Pros: GC functions partially even when not all sources are ready
Cons: unclean container runtime interface.

Random-Liu commented:
Vote for option 2, which is cleaner than current PR and still support garbage collect before all sources are ready.

yujuhong committed:
Don't remove all containers of deleted pods until sources are ready

Without this fix, kubelet may assume a pod from a not-ready source has already
been deleted, and GC all its dead containers.

Random-Liu commented:
@yujuhong Have you decided to go with option 1? If so, LGTM.

yujuhong commented:
Let's go with option 2

yujuhong commented:
/cc @dchen1107

ronnielai commented on pkg/kubelet/container/container_gc.go:
```diff
@@ -39,7 +39,7 @@ type ContainerGCPolicy struct {
 type ContainerGC interface {
 	// Garbage collect containers.
-	GarbageCollect() error
+	GarbageCollect(allSourcesReady bool) error
```
Using boolean parameters in a public method is generally frowned upon. Just my 2 cents.

Random-Liu replied:
@ronnielai This is just a quick fix for this. :)
This is definitely not what we want, but we are in code freeze, we've to fix it with as few code as possible. :)

yujuhong replied:
We don't want to refactor the code extensively due to reason that @Random-Liu said. Making this part of the ContainerGCPolicy is even uncleaner, since the state changes dynamically (at least during kubelet startup).

Random-Liu replied:
@yujuhong Can we add a few comments here to explain why we add this and when are we going to remove it? Or link an issue.

Random-Liu commented:
@yujuhong 
```
FAIL	k8s.io/kubernetes/pkg/kubelet/rkt [build failed]
```

Random-Liu commented:
This time it's docker:
```
FAIL	k8s.io/kubernetes/pkg/kubelet/dockertools [build failed]
```

yujuhong commented:
Oops....sorry. I expected the log to show me all the failures, but it didn't. Ran the unit test locally and passed.

Random-Liu commented:
LGTM with a nit.

yujuhong commented:
Done.

Random-Liu commented:
@yujuhong Maybe we need some labels to mark this should be backported in v1.2?

yujuhong commented:
@dchen1107, do we still accept cherry picks in v1.2?

dchen1107 commented:
We still accept the cherry picks in v1.2, but I don't think we need to cherrypick the fix for this particular issue since the issue only affects static pods with RestartPolicy:never.

yifan-gu commented:
rkt changes LGTM

lavalamp commented:
manual build cop merge
